grammar.js

```diff
@@ -205,46 +205,7 @@ module.exports = grammar({
       seq("->", repeat1(alias($._decimal_integer, $.integer))),
 
     asm_function_body: ($) =>
-      seq(
-        "{",
-        prec.right(
-          repeat(
-            choice(
-              // list with { }
-              $.asm_list,
-              // others
-              $._asm_instruction,
-            ),
-          ),
-        ),
-        prec.right("}"),
-      ),
-
-    asm_list: ($) => seq("{", /\s/, repeat($._asm_instruction), "}", /\s/),
-
-    _asm_instruction: ($) =>
-      choice(
-        // string
-        $._asm_string,
-        // char
-        seq("char", /\s/, /\S/, /\s/),
-        // custom
-        /\S+/, // NOTE: this point can be significantly improved
-      ),
-
-    _asm_string: (_) =>
-      seq(
-        choice('abort"', '."', '+"', '"'),
-        token.immediate(prec(1, /[^"]+/)),
-        token.immediate('"'),
-        /\s/,
-      ),
-
-    // NOTE: May be re-introduced in the future, unused in the current parser
-    // listNoStateCheck
-    // seq("({)", /\s/, repeat($._asm_instruction), "(})", /\s/),
-    // hexLiteral
-    // _asm_hex_literal: (_) => /[xB]\{[\s\da-fA-F]*_?\s*\}\s/,
+      seq("{", prec.right(repeat($._tvm_statement)), prec.right("}")),
```

Comment:
We struggled a bit with what can be done with assembly and ended up with this grammar.

https://github.com/tact-lang/tact/blob/main/src/grammar/next/grammar.gg#L162-L168

The thing that worries me is that these two grammars do not match.

Reply:
Ah, interesting, so the following code is also valid?

```
"string item"
{
    "nested string"
    // Another comment
}
"more strings"
```

```
any text here
```

```
// just comment
```

Looks like the compiler's grammar is even more loose than this (no numbers for example) 🤔

Reply:
I don't think it's a big problem, we need more precise grammar in tree-sitter for better highlighting and tools like Language Server or something like that

Reply:
Looks like it's time to start designing the real grammar for TVM assembly (we will need it in our new backend)

Reply:
> Looks like it's time to start designing the real grammar for TVM assembly (we will need it in our new backend)

Yeah, we need it for the JetBrains plugin too.

Reply:
@i582 Fift doesn't have a grammar. It can redefine any input tokens to mean anything, and even have mismatched brackets. To get a syntax highlighting for it, or even parse correctly Tact that includes it, we have to interpret it.

Until there's no proper TVM assembler in Tact, we want to just pass majority of Fift programs through.

Reply:
> Until there's no proper TVM assembler in Tact, we want to just pass majority of Fift programs through.

Got it, looks like I need to revert changes to fift stage :D

Reply:
> Or we can just wait for proper support in Tact and then I'll replicate it here

Yup, which I do with all Tact releases :)
And for tact.vim and tact-sublime too.

Reply:
@i582 If you want to have the closest tree-sitter to the actual upstream grammar of Tact, just let me know, and I'll update the #45 branch ahead of time.